### Import required dependencies

In [1]:
import pandas as pd
import os
import numpy as np

## Deliverable 1: Collect the Data

To collect the data that you’ll need, complete the following steps:

1. Using the Pandas `read_csv` function and the `os` module, import the data from the `new_full_student_data.csv` file, and create a DataFrame called student_df. 

2. Use the head function to confirm that Pandas properly imported the data.


In [4]:
# Create the path and import the data
new_full_student_data = os.path.join('../Resources/new_full_student_data.csv')
student_df = pd.read_csv(new_full_student_data)
student_df

student_id      student_name grade             school_name  \
0       103880842     Travis Martin   9th    Sullivan High School   
1        45069750     Michael Brown   9th       Dixon High School   
2        45024902   Gabriela Lucero   9th      Wagner High School   
3        62582498  Susan Richardson   9th       Silva High School   
4        16437227      Sherry Davis  11th      Bowers High School   
...           ...               ...   ...                     ...   
19509   109236636     Robert Sawyer  12th       Silva High School   
19510    63239258      David Herman   9th       Woods High School   
19511    95516554        Megan Gill  11th      Wagner High School   
19512    65050383        Lori Stone  11th      Bowers High School   
19513    34720657       Anna Jensen  12th  Montgomery High School   

       reading_score  math_score school_type  school_budget  
0               59.0        88.2      Public         961125  
1               94.7        73.5     Charter         870334  
2               89.0        70.4      Public         846745  
3               69.7        80.3      Public         991918  
4                NaN        27.5      Public         848324  
...              ...         ...         ...            ...  
19509           43.3        27.2      Public         991918  
19510           52.1        80.4      Public         912243  
19511           93.9        84.1      Public         846745  
19512           94.6        70.9      Public         848324  
19513           82.3        42.4     Charter         893368  

[19514 rows x 8 columns]

In [5]:
# Verify that the data was properly imported
student_df.head()

student_id      student_name grade           school_name  reading_score  \
0   103880842     Travis Martin   9th  Sullivan High School           59.0   
1    45069750     Michael Brown   9th     Dixon High School           94.7   
2    45024902   Gabriela Lucero   9th    Wagner High School           89.0   
3    62582498  Susan Richardson   9th     Silva High School           69.7   
4    16437227      Sherry Davis  11th    Bowers High School            NaN   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
4        27.5      Public         848324

## Deliverable 2: Prepare the Data

To prepare and clean your data for analysis, complete the following steps:
    
1. Check for and remove all rows with `NaN`, or missing, values in the student DataFrame. 

2. Check for and remove all duplicate rows in the student DataFrame.

3. Use the `str.replace` function to remove the "th" from the grade levels in the grade column.

4. Check data types using the `dtypes` property.

5. Remove the "th" suffix from every value in the grade column using `str` and `replace`.

6. Change the grade colum to the `int` type and verify column types.

7. Use the head (and/or the tail) function to preview the DataFrame.

In [6]:
# Check for NaN values
student_df.isna()

student_id  student_name  grade  school_name  reading_score  \
0           False         False  False        False          False   
1           False         False  False        False          False   
2           False         False  False        False          False   
3           False         False  False        False          False   
4           False         False  False        False           True   
...           ...           ...    ...          ...            ...   
19509       False         False  False        False          False   
19510       False         False  False        False          False   
19511       False         False  False        False          False   
19512       False         False  False        False          False   
19513       False         False  False        False          False   

       math_score  school_type  school_budget  
0           False        False          False  
1           False        False          False  
2           False        False          False  
3           False        False          False  
4           False        False          False  
...           ...          ...            ...  
19509       False        False          False  
19510       False        False          False  
19511       False        False          False  
19512       False        False          False  
19513       False        False          False  

[19514 rows x 8 columns]

In [7]:
# Check and count for all NaN values
student_df.isna().sum()

student_id          0
student_name        0
grade               0
school_name         0
reading_score    1968
math_score        982
school_type         0
school_budget       0
dtype: int64

In [8]:
# Drop rows with null values and verify removal
student_df = student_df.dropna()
student_df.isna().sum()

student_id       0
student_name     0
grade            0
school_name      0
reading_score    0
math_score       0
school_type      0
school_budget    0
dtype: int64

In [9]:
# Check for duplicated rows
student_df.duplicated()

0        False
1        False
2        False
3        False
5        False
         ...  
19509    False
19510     True
19511    False
19512    False
19513    False
Length: 16667, dtype: bool

In [10]:
#Count how many duplicated values exist in the data
student_df.duplicated().sum()

1836

In [11]:
# Drop duplicated rows and verify removal
student_df.drop_duplicates()
student_df=student_df.drop_duplicates()
student_df.duplicated().sum()

0

In [12]:
# Check data types
student_df.dtypes

student_id         int64
student_name      object
grade             object
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [13]:
# Examine the grade column to understand why it is not an int
student_df['grade']

0         9th
1         9th
2         9th
3         9th
5         9th
         ... 
19508    10th
19509    12th
19511    11th
19512    11th
19513    12th
Name: grade, Length: 14831, dtype: object

In [14]:
# Remove the non-numeric characters and verify the contents of the column
student_df['grade']=student_df['grade'].str.replace("th","")
student_df['grade']

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: object

In [15]:
# Change the grade column to the int type and verify column types
student_df ['grade']=student_df['grade'].astype(int)
student_df.dtypes

student_id         int64
student_name      object
grade              int32
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [16]:
student_df

student_id      student_name  grade             school_name  \
0       103880842     Travis Martin      9    Sullivan High School   
1        45069750     Michael Brown      9       Dixon High School   
2        45024902   Gabriela Lucero      9      Wagner High School   
3        62582498  Susan Richardson      9       Silva High School   
5        74579444   Cynthia Johnson      9  Montgomery High School   
...           ...               ...    ...                     ...   
19508    83985333   Deborah Sanders     10       Silva High School   
19509   109236636     Robert Sawyer     12       Silva High School   
19511    95516554        Megan Gill     11      Wagner High School   
19512    65050383        Lori Stone     11      Bowers High School   
19513    34720657       Anna Jensen     12  Montgomery High School   

       reading_score  math_score school_type  school_budget  
0               59.0        88.2      Public         961125  
1               94.7        73.5     Charter         870334  
2               89.0        70.4      Public         846745  
3               69.7        80.3      Public         991918  
5               63.5        76.9     Charter         893368  
...              ...         ...         ...            ...  
19508           60.5        64.6      Public         991918  
19509           43.3        27.2      Public         991918  
19511           93.9        84.1      Public         846745  
19512           94.6        70.9      Public         848324  
19513           82.3        42.4     Charter         893368  

[14831 rows x 8 columns]

## Deliverable 3: Summarize the Data

Describe the data using summary statistics on the data as a whole and on individual columns.

1. Generate the summary statistics for each DataFrame by using the `describe` function.

2. Display the mean math score using the `mean` function. 

2. Store the minimum reading score as `min_reading_score`.

In [17]:
# Display summary statistics for the DataFrame
student_df.describe()

student_id         grade  reading_score    math_score  school_budget
count  1.483100e+04  14831.000000   14831.000000  14831.000000   14831.000000
mean   6.975296e+07     10.355539      72.357865     64.675733  893742.749107
std    3.452909e+07      1.097728      15.224590     15.844093   53938.066467
min    1.000906e+07      9.000000      10.500000      3.700000  817615.000000
25%    3.984433e+07      9.000000      62.200000     54.500000  846745.000000
50%    6.965978e+07     10.000000      73.800000     65.300000  893368.000000
75%    9.927449e+07     11.000000      84.000000     76.000000  956438.000000
max    1.299997e+08     12.000000     100.000000    100.000000  991918.000000

In [18]:
# Display the mean math score using the mean function
student_df.mean()

C:\Users\rdono\AppData\Local\Temp\ipykernel_32956\969459584.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  student_df.mean()


student_id       6.975296e+07
grade            1.035554e+01
reading_score    7.235787e+01
math_score       6.467573e+01
school_budget    8.937427e+05
dtype: float64

In [19]:
student_df['math_score'].mean()

64.67573326141189

In [20]:
# Store the minimum reading score as min_reading_score
student_df['math_score'].min()

3.7

## Deliverable 4: Drill Down into the Data

Drill down to specific rows, columns, and subsets of the data.

To drill down into the data, complete the following steps:

1. Use `loc` to display the grade column.

2. Use `iloc` to display the first 3 rows and columns 3, 4, and 5.

3. Show the rows for grade nine using `loc`.

4. Store the row with the minimum overall reading score as `min_reading_row` using `loc` and the `min_reading_score` found in Deliverable 3.

5. Find the reading scores for the school and grade from the output of step three using `loc` with multiple conditional statements.

6. Using conditional statements and `loc` or `iloc`, find the mean reading score for all students in grades 11 and 12 combined.

In [21]:
student_df.loc[:,"grade"]

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: int32

In [22]:
# Use `iloc` to display the first 3 rows and columns 3, 4, and 5.
student_df.iloc[0:3,[3,4,5]]

school_name  reading_score  math_score
0  Sullivan High School           59.0        88.2
1     Dixon High School           94.7        73.5
2    Wagner High School           89.0        70.4

In [23]:
# Select the rows for grade nine and display their summary statistics using `loc` and `describe`.
grade9_student_df = student_df.loc[student_df["grade"] == 9]
grade9_student_df.describe()

student_id   grade  reading_score   math_score  school_budget
count  4.132000e+03  4132.0    4132.000000  4132.000000    4132.000000
mean   6.979441e+07     9.0      69.236713    66.585624  898692.606002
std    3.470565e+07     0.0      15.277354    16.661533   54891.596611
min    1.000906e+07     9.0      17.900000     5.300000  817615.000000
25%    3.953848e+07     9.0      59.000000    56.000000  846745.000000
50%    6.984037e+07     9.0      70.050000    67.800000  893368.000000
75%    9.939504e+07     9.0      80.500000    78.500000  957299.000000
max    1.299997e+08     9.0      99.900000   100.000000  991918.000000

In [24]:
# Store the row with the minimum overall reading score as `min_reading_row`
# using `loc` and the `min_reading_score` found in Deliverable 3.
min_reading_score= student_df["reading_score"].min()
min_reading_score_row = student_df.loc[student_df["reading_score"]==min_reading_score]
min_reading_score_row

student_id    student_name  grade        school_name  reading_score  \
3706    81758630  Matthew Thomas     10  Dixon High School           10.5   

      math_score school_type  school_budget  
3706        58.4     Charter         870334

In [25]:
# Use loc with conditionals to select all reading scores from 10th graders at Dixon High School.
student_df.loc[student_df["grade"] == 10, ["school_name","reading_score"]]
student_df.loc[(student_df["school_name"]== "Dixon High School") & (student_df["grade"]==10), ["school_name","reading_score"]]


school_name  reading_score
45     Dixon High School           71.1
60     Dixon High School           59.5
69     Dixon High School           88.6
94     Dixon High School           81.5
100    Dixon High School           95.3
...                  ...            ...
19283  Dixon High School           52.9
19306  Dixon High School           58.0
19344  Dixon High School           38.0
19368  Dixon High School           84.4
19445  Dixon High School           43.9

[569 rows x 2 columns]

In [26]:
# Find the mean reading score for all students in grades 11 and 12 combined.
student_df.loc[(student_df["grade"] == 11) | (student_df["grade"] == 12), ["reading_score"]].mean()

reading_score    74.900381
dtype: float64

## Deliverable 5: Make Comparisons Between District and Charter Schools

Compare district vs charter schools for budget, size, and scores.

Make comparisons within your data by completing the following steps:


In [27]:
# Use groupby and mean to find the average budget for school type.
school_type_budget_df = student_df.groupby(["school_type"]).mean()
school_type_budget_df.loc[:,["school_budget"]]

school_budget
school_type               
Charter      872625.656236
Public       911195.558251

In [29]:
# Use the `groupby`, `count`, and `sort_values` functions to find the
# total number of students at each school and sort from most students to least students.
student_count_df = student_df.groupby(["school_name"]).count()
student_df.rename(columns = {"student_id":"student_count"}, inplace = True)
student_count_df.loc[:,["student_count"]].sort_values("student_count",ascending=False)

student_count
school_name                          
Montgomery High School           2038
Green High School                1961
Dixon High School                1583
Wagner High School               1541
Silva High School                1109
Woods High School                1052
Sullivan High School              971
Turner High School                846
Bowers High School                803
Fisher High School                798
Richard High School               551
Campos High School                541
Odonnell High School              459
Campbell High School              407
Chang High School                 171

In [30]:
# Find the average math score by grade for each school type using groupby and mean functions
school_math_score_df = student_df.groupby(["school_type","grade"]).mean()
school_math_score_df.loc[:,["math_score"]]

math_score
school_type grade            
Charter     9       70.077874
            10      66.443206
            11      68.024735
            12      60.212121
Public      9       63.771066
            10      63.764121
            11      59.314337
            12      63.568319

# Deliverable 6: Summarize Your Findings
In the cell below, write a few sentences to describe any discoveries you made while performing your analysis along with any additional analysis you believe would be worthwhile.

***Summary here***

This project consisted in analyzing data on school budget, school size and student funding for standardized test scores (math and reading tests). 

During this process we found 2950 missing records (1968 regarding reading scores and 982 for math scores). Furthermore, we found a total of 1836 duplicated entries. To avoid skew data sums and averages by inflating numbers or create other problems during an analysis, we dropped all null and duplicated values. After dropping all (null and duplicated) values, we have come up to the conclusion that these results in both, Public and Charter schools, are pretty close regardless of their type and budget, as well as number of students (size).

In addition to that, we performed 5 more analysis to validate our results (reading_score (mean), max_math_score, min_math_score, max_reading_score, group by school size - type), it is clear that results are relatively similar despite of school budget, school size and school type. In other words, out of the 12 best score tests in both math and reading 5 belong to public schools and 7 belong to charter schools. 

Finally, it’d be worth mentioning that 3 of the best scores belong to Dixon High School.  


In [31]:
school_reading_score_df = student_df.groupby(["school_type","grade"]).mean()
school_reading_score_df.loc[:,["reading_score"]]

reading_score
school_type grade               
Charter     9          68.676790
            10         69.628822
            11         80.596379
            12         70.475421
Public      9          69.687981
            10         73.160461
            11         73.486448
            12         73.342380

In [32]:
max_reading_score= student_df["reading_score"].max()
max_reading_score_row = student_df.loc[student_df["reading_score"]==max_reading_score]
max_reading_score_row

student_count    student_name  grade         school_name  \
3605        22589871    Jeremy Wells     11   Dixon High School   
5041        39568500  Robert Johnson     12   Dixon High School   
11813       63851186       Mary Rush     11   Silva High School   
17576       45424751      Robert Kim     12  Wagner High School   
17908       47461327   Morgan Vaughn     11  Wagner High School   

       reading_score  math_score school_type  school_budget  
3605           100.0        91.7     Charter         870334  
5041           100.0        73.9     Charter         870334  
11813          100.0        63.8      Public         991918  
17576          100.0        72.8      Public         846745  
17908          100.0        61.4      Public         846745

In [33]:
max_math_score= student_df["math_score"].max()
max_math_score_row = student_df.loc[student_df["math_score"]==max_math_score]
max_math_score_row

student_count        student_name  grade             school_name  \
2344        19976785        Allen Bender     10    Campbell High School   
4247       103453606  Jennifer Frederick     11  Montgomery High School   
5467        49587791       Duane Jackson      9       Green High School   
8031        50675621      Jennifer Jones     10       Dixon High School   
14125       66107992           Jon Smith     10    Campbell High School   
18949       94926637   Jeffrey Hernandez      9    Sullivan High School   
19327       32931286      Kristina Baker      9      Fisher High School   

       reading_score  math_score school_type  school_budget  
2344            75.6       100.0     Charter         960726  
4247            90.7       100.0     Charter         893368  
5467            75.8       100.0     Charter         832670  
8031            55.1       100.0     Charter         870334  
14125           49.7       100.0     Charter         960726  
18949           74.7       100.0      Public         961125  
19327           78.7       100.0      Public         817615

In [34]:
# Store the row with the minimum overall reading score as `min_reading_row`
# using `loc` and the `min_reading_score` found in Deliverable 3.
min_math_score= student_df["math_score"].min()
min_math_score_row = student_df.loc[student_df["math_score"]==min_math_score]
min_math_score_row

student_count student_name  grade             school_name  \
10419      123372576  Nicole Soto     10  Montgomery High School   

       reading_score  math_score school_type  school_budget  
10419           75.5         3.7     Charter         893368

In [41]:
# Use the `groupby`, `count`, and `sort_values` functions to find the
# total number of students at each school and sort from most students to least students.
student_count_df = student_df.groupby(["school_type", "school_name"]).count()
student_df.rename(columns = {"student_id":"student_count"}, inplace = True)
student_count_df.loc[:,["student_count"]].sort_values("student_count",ascending=False)

student_count
school_type school_name                          
Charter     Montgomery High School           2038
            Green High School                1961
            Dixon High School                1583
Public      Wagner High School               1541
            Silva High School                1109
            Woods High School                1052
            Sullivan High School              971
            Turner High School                846
            Bowers High School                803
            Fisher High School                798
Charter     Richard High School               551
Public      Campos High School                541
            Odonnell High School              459
Charter     Campbell High School              407
            Chang High School                 171